operations/observability/Observability.md

```diff
@@ -0,0 +1,336 @@
+
+Goss test logs are scraped using grok-exporter and visualized on Grafana using captured data.
+
+Goss tests are contained in YAML file. A YAML file containing a list of Goss test files is called a test suite. These suites are run through the CSM-provided scripts to get a log file with results for the tests.
+
+### Workflow
+
+![Goss test workflow](../../img/operations/GossWorkflow.png "Goss Workflow")
+
+### Log file format
+
+Individual lines of logs are in the following format for each node, test name, and source:
+
+```json
+{
+  "log_timestamp": "20230118_094205.821955",
+  "Product": "CSM",
+  "log_script": "print_goss_json_results.py",
+  "log_message": "Test result",
+  "Description": "Validates that 'cray --version' is available and executes successfully on the local system.",
+  "Execution Time (nanoseconds)": 1312368478,
+  "Execution Time (seconds)": 1.312368478,
+  "Node": "ncn-m001",
+  "Result Code": 0,
+  "Result String": "PASS",
+  "Source": "http://ncn-m001.hmn:8997/ncn-healthcheck-master",
+  "Test Name": "Command 'cray --version' Available",
+  "Test Summary": "Command: command_available_1_cray_version: exit-status: matches expectation: [0]"
+}
+```
+
+### Grok-exporter deployment, service, and service-monitor
+
+Grok-exporter is deployed on all of the Kubernetes master nodes, by using node affinity in its deployment.
+As background information, the service for the grok-exporter is accessible at port `9144` and service-monitor implementation is for Prometheus to access the metrics that are created by the grok-exporter.
+
+### Configuration file for the grok-exporter
+
+The configuration file for the grok-exporter provides instructions on how to parse the Goss test log files and creates metrics from them.
+grok-exporter version 3 configuration is used for this task and to match log expressions with regular expressions.
+
+The following is an example configuration passed to the grok-exporter to get metrics:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: example_name
+  namespace: example_namespace
+data:
+  config.yml: |-
+    global:
+      config_version: 3
+    input:
+      type: file
+      paths: 
+      - /logs/goss_tests/*
+      fail_on_missing_logfile: false
+    grok_patterns:
+      - 'REGEX [regular].*expression'
+      - 'EXAMPLE [aA-zZ-].*'
+    metrics:
+      - type: gauge
+        name: example_metric
+        help: EXAMPLE METRIC HELP
+        match: '{"log_file": "%{REGEX:logfile}", "TEST": "%{EXAMPLE:ex}"}'
+        value: '{{`{{.logfile}}`}}'
+        labels:
+          example_test: '{{`{{.ex}}`}}'
+    server:
+     port: 9144
+```
+
+### Prometheus metrics and Grafana dashboard
+
+After the preceding steps the `goss_tests` metrics are seen in Prometheus when Goss tests are run. Using these metrics,
+Grafana dashboards are created to shows the Goss tests details visually.
```

Comment:
Once configured as described in the preceding section, the `goss_tests` metrics are displayed in Prometheus whenever Goss tests are run.  Grafana dashboards are created and use these metrics to display the Goss test details visually.